cmd/cli/config/set.go

```diff
@@ -0,0 +1,160 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/bacalhau-project/bacalhau/pkg/config"
+	"github.com/bacalhau-project/bacalhau/pkg/config/types"
+	"github.com/bacalhau-project/bacalhau/pkg/logger"
+	"github.com/bacalhau-project/bacalhau/pkg/model"
+)
+
+func newSetCmd() *cobra.Command {
+	showCmd := &cobra.Command{
+		Use:   "set",
+		Args:  cobra.MinimumNArgs(2),
+		Short: "Set a value in the config.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return setConfig(args[0], args[1:]...)
+		},
+	}
+	return showCmd
+}
+
+func setConfig(key string, values ...string) error {
+	// there may me a config file present, we'll write to that if it exists.
+	configFile := viper.ConfigFileUsed()
+
+	// we create a new viper instance that we'll use to set values on and update the config.
+	viperWriter := viper.New()
+	viperWriter.SetTypeByDefaultValue(true)
+	if configFile == "" {
+		// if there isn't a config file, we'll assume we add it to the current repo.
+		configFile = filepath.Join(viper.GetString("repo"), "config.yaml")
+	}
+	viperWriter.SetConfigFile(configFile)
+	// the instance has read in a copy of the config file
+	if err := viperWriter.ReadInConfig(); err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+	}
+
+	// we create a viper instance to serve as a hacky schema
+	viperSchema := viper.New()
+	// we load in the default configuration to our viper schema instance
+	types.SetDefaults(config.ForEnvironment(), types.WithViper(viperSchema))
+	//calling `Get` on this instance will return a default value from the config structure that we can type assert on.
+	curValue := viperSchema.Get(key)
+
+	// we need special handling for the nodetype since its just a string, but carries implicit requirements in the config.
+	if strings.EqualFold(types.NodeType, key) {
+		if !validNodeTypes(values) {
+			return fmt.Errorf("setting: %q, invalid node type value: %q, must be one of: 'requester' 'compute' 'requester compute'", key, values)
+		}
+	}
+	switch curValue.(type) {
```

Comment:
This bit of code is why the code gen changed - needed multiple viper instances to make this work.

Reply:
Agreed, a data-first approach would be better here and would avoid a lot of the repetition. We even already have some of this available as flags, which could be adapted to work in both places. https://github.com/bacalhau-project/bacalhau/blob/bfc436a9afb4b62134321f390a94883130ed39c8/cmd/util/flags/types.go#L296-L303

Reply:
So something like this?
```
type Parser[T any] func(...string) (T, error)

type ValueParser[T any] struct {
	parser Parser[T]
}
```
```go
	case types.StorageType:
		vp := &ValueParser[types.StorageType]{
			parser: func(values ...string) (types.StorageType, error) {
				value, err := singleValueOrError(values...)
				if err != nil {
					return 0, err
				}

				var out types.StorageType
				if err := flags.StorageTypeFlag(&out).Set(value); err != nil {
					return 0, err
				}
				return out, nil
			},
		}
```
then later after the switch:
```go
		if val, err := vp.parser(values...); err != nil {
			return err
		} else {
			viperWriter.Set(key, val)
		}
```
I think this might save us 10-15LOC at the cost of greater cognitive overhead. @rossjones how strongly do you feel about trying to shorten the `switch`? If you have a better example than than this I'd be happy to implement it, but based on what I tried this doesn't feel simpler, just shorter.

Reply:
@simonwo 
> We even already have some of this available as flags, which could be adapted to work in both places

Love it, adapting to use these

Reply:
> I think this might save us 10-15LOC at the cost of greater cognitive overhead.

I sorta meant something like this which saves 60+ lines and I think might be cognitively simpler if you think of it in terms of "what parser do I need for this type"?

```go
(string) (any, error)
	var parser parserFunc

	switch curValue.(type) {

	case string:
		parser = func(v string) (any, error) { return v, nil } // identity by default
	case bool:
		parser = func(v string) (any, error) { return strconv.ParseBool(v) }
	case int, int8, int16, int32, int64:
		parser = func(v string) (any, error) { return strconv.ParseInt(v, 10, 64) }
	case float32, float64:
		parser = func(v string) (any, error) { return strconv.ParseFloat(v, 64) }
	case types.Duration:
		parser = func(v string) (any, error) { return time.ParseDuration(v) }
	case model.JobSelectionDataLocality:
		parser = func(v string) (any, error) { model.ParseJobSelectionDataLocality(v) }
	case logger.LogMode:
		parser = func(v string) (any, error) { logger.ParseLogMode(v) }
	case types.StorageType:
		parser = func(v string) (any, error) { types.ParseStorageType(v) }
	default:
		return fmt.Errorf("unsupported type %T for key: %q", curValue, key)
	}

	value, err := singleValueOrError(values...)
	if err != nil {
		return fmt.Errorf("setting %q: %w", key, err)
	}

	configValue, err := parser(value)
	if err != nil {
		return fmt.Errorf("setting %q: invalid boolean value: %q: %w", key, value, err)
	}
	viperWriter.Set(key, configValue)
```

Reply:
I see, I've incorporated this change.